<img src="img/datenschule_logo_web.png" width="25%"  align="left">

<img src="img/coat_of_arms_of_saxony.svg" width="12%"  align="right">
# Aktivitäten und Partnerschaften an Schulen in Sachsen
In diesem Notebook dokumnetieren wir unser Vorgehen bei der Auswertung der Partnerschaften und Aktivitäten <br> der Schulen in Sachsen. Da wir die Daten auf der Website laufend aktualisiert, können die Ergebnisse hier von denen <br> auf der Website abweichen. Der Berechnungsweg ist allerdings der Gleiche.<br>
Mehr Informationen über das Schulsystem in Deutschland gibt es auf <a href='https://jedeschule.de'>jedeschule.de<a>

In [1]:
#Für die Berechnung nutzen wir hier die Pandas-Library
import pandas as pd
import urllib.request, json

In [2]:
with urllib.request.urlopen("https://lab.okfn.de/jedeschule/all_schools_geocoded.json") as url:
    data = json.loads(url.read().decode())

In [20]:
sachsen_data = [item for item in data if item['state'] == 'SN']

[{'address': ' \xa0 04229\xa0Leipzig\xa0OT\xa0Zentrum  ',
  'full_time_school': False,
  'id': 'SN-4490551-0',
  'lat': 51.333713,
  'legal_status': 1,
  'lon': 12.319407,
  'name': 'Förderpädagogische Beratungsstelle (Abordnungsdienststelle der SBA Leipzig)',
  'official_id': '4490551',
  'partner': [],
  'profile': True,
  'programs': {'programs': [], 'working_groups': []},
  'provider': 'Freistaat Sachsen - Staatsministerium für Kultus',
  'school_id': '4490551',
  'school_type': 'Sonstige Bildungs- und Betreuungseinrichtungen Förderpädagogische Beratungsstelle',
  'school_type_entity': 'Sonstige Bildungs- und Betreuungseinrichtungen Förderpädagogische Beratungsstelle',
  'state': 'SN',
  'students': {'2013': [], '2014': [], '2015': []},
  'teacher': []},
 {'address': ' Wiener Platz\xa06 01069\xa0Dresden\xa0OT\xa0Südvorstadt-West  ',
  'email': 'info@es.dresden.eso.de',
  'full_time_school': False,
  'id': 'SN-4390688-0',
  'lat': 51.041984,
  'legal_status': 1,
  'lon': 13.731929,


In [4]:
partner_data = []
wg_data = []
for school in sachsen_data:
    for partner in school['partner']:
        partner_data.append({
            "Schule": school['name'],
            "ID": school['id'],
            "school_type" : school['school_type'],
            "partner_name" : partner['name'],
            "Type_grob": partner['type']['grob']#,
#             "Type_mittel": partner['type']['mittel'],
#             "Type_detail": partner['type']['detail']
        })
    for wg in school['programs']['working_groups']:
        wg_data.append({
            "id": school['id'],
            "ag_name" : wg['entity'],
            "ag_cat" : wg['category'],
            "school" : school['official_id']
        })


In [16]:
sachsen_df_partner = pd.DataFrame.from_dict(partner_data)
sachsen_partner_df.head(5)

ID                                             Schule  \
0  SN-4220216-0  Sonderpädagogisches Förderzentrum  "Johannes T...   
1  SN-4220216-0  Sonderpädagogisches Förderzentrum  "Johannes T...   
2  SN-4220216-0  Sonderpädagogisches Förderzentrum  "Johannes T...   
3  SN-4520102-0    Alte Reusaer Schule, Schule für Erziehungshilfe   
4  SN-4520243-0      Sperlingsbergschule, Schule zur Lernförderung   

                   Type_grob  \
0      Gemeinnütziger Akteur   
1                 Unbestimmt   
2          Wirtschaftsakteur   
3  Öffentliche Infrastruktur   
4  Öffentliche Infrastruktur   

                                       partner_name  \
0           Kinder-, Jugend- und Familienhilfe e.V.   
1          diverse Mittelschulen der Stadt Chemnitz   
2  Solaris Förderzentrum für Jugend und Umwelt GmbH   
3             Jugendamt Stadt Plauen, Vogtlandkreis   
4                       Daetz- Zentrum Lichtenstein   

                  school_type  
0  Schule für Erziehungshilfe  
1  Schule für Erziehungshilfe  
2  Schule für Erziehungshilfe  
3  Schule für Erziehungshilfe  
4    Schule zur Lernförderung

In [6]:
sachsen_ag_df = pd.DataFrame.from_dict(wg_data)
sachsen_ag_df.head(5)

ag_cat                    ag_name            id  \
0  Gesellschaft / Partizipation        Gesellschaftsspiele  SN-4220216-0   
1  Gesellschaft / Partizipation                Schülerclub  SN-4220216-0   
2            Literatur / Medien          Schülerbibliothek  SN-4220216-0   
3            Berufsorientierung       Offenes Spielangebot  SN-4220216-0   
4                         Sport  Allgemeiner Freizeitsport  SN-4220216-0   

    school  
0  4220216  
1  4220216  
2  4220216  
3  4220216  
4  4220216

### 1. Auswertung der Schulaktivitäten an Schulen in Sachsen

In [7]:
df_sachsen = pd.read_csv('data/sachsen/ag_sachsen.csv', sep=',')

Zunächst werden die CSV-Datei zu den Aktivitäten an Schulen in Sachsen importiert. Die Daten liegen in unserem Github Ordner.

In [8]:
df_sachsen.head(5)

id                    ag_name                        ag_cat  \
0  SN-4220216-0        Gesellschaftsspiele  Gesellschaft / Partizipation   
1  SN-4220216-0                Schülerclub  Gesellschaft / Partizipation   
2  SN-4220216-0          Schülerbibliothek            Literatur / Medien   
3  SN-4220216-0       Offenes Spielangebot            Berufsorientierung   
4  SN-4220216-0  Allgemeiner Freizeitsport                         Sport   

    school  
0  4220216  
1  4220216  
2  4220216  
3  4220216  
4  4220216

Zu Beginn einer Analyse ist es sinnvoll, einen Überblick über die Daten zu gewinnen.

In [9]:
df_sachsen.drop_duplicates().describe(include='all')

id ag_name ag_cat        school
count          11633   11633  11633  1.163300e+04
unique          1206     164     10           NaN
top     SN-4540340-0    Chor  Sport           NaN
freq              50     579   3203           NaN
mean             NaN     NaN    NaN  4.317592e+06
std              NaN     NaN    NaN  1.324859e+05
min              NaN     NaN    NaN  4.050019e+06
25%              NaN     NaN    NaN  4.212411e+06
50%              NaN     NaN    NaN  4.312393e+06
75%              NaN     NaN    NaN  4.420303e+06
max              NaN     NaN    NaN  4.550778e+06

Mit dem Befehl *'describe()'* können erhalten wir eine Zusammenfassung der Daten. Wir sehen:
unique: 1206 der 1.500 allgemeinbildenden Schulen in Sachsen kommunizieren ihre Projekte und Aktivitäten an das Landesamt Sachsen.
count: Es werden über 11.600 Aktivitäten an Schulen in Sachsen angeboten.

In [11]:
#Die Anzahl der Schule in Sachsen
schools_count = len(df_sachsen.school.unique())

schooltypes_count = len(df_sachsen.id.unique())

# Anzahl der angebotenen Aktivitäten
projects_count = len(df_sachsen)

# Durchschnittliche Anzahl an Projekten an Schulen in Sachsen
projects_avg = projects_count / schooltypes_count

print('Eine Schule in Sachsen bietet ihren Schüler*innen durchschnittlich %s Aktivitäten an.' %(round(projects_avg,2)))

Eine Schule in Sachsen bietet ihren Schüler*innen durchschnittlich 11.61 Aktivitäten an.


Um zu errechnen wie viele Projekte eine Schule in Sachsen durchschnittlich anbietet, teilen wir die Anzahl der Projekte durch die Anzahl der Schulen. Da die CSV-Datei für jedes Angebot einen eigenen Zeileneintrag enthält, werden auch Schulen doppelt genannt. Mit der Funktion *'unique()'* filtern wir doppelte Schuleinträge aus. Dadurch erhalten wir die Anzahl der Schulen in Sachsen, die ihre Aktivitäten kommuniziert haben.
**Ergebnis:**
* Durchschnittlich bietet eine Schule in Sachsen ihren Schüler*innen knapp 12 Aktivitäten an.

In [12]:
#Wie viel Projekte aus unterschiedlichen Bereichen bieten die Schulen ihren Schüler*innen?
category_unique = df_sachsen.drop_duplicates(subset=['id','ag_cat'])

#Calculate average number of Activity-Categories per School
category_avg = len(category_unique) / schooltypes_count
print('In Sachsen werden in durchschnittlich %s Kategorien Aktivitäten angeboten.' %(round(category_avg,2)))

In Sachsen werden in durchschnittlich 5.24 Kategorien Aktivitäten angeboten.


In [13]:
categories_pivot = pd.pivot_table(category_unique,index=['ag_cat'],values='id',aggfunc='count')
categories = pd.DataFrame(categories_pivot)
categories['% Schulen mit Projekt-Kategorie']= (categories.id / schooltypes_count * 100).round(2)
categories.sort_values(by='id',ascending=False)

id  % Schulen mit Projekt-Kategorie
ag_cat                                                             
Sport                         1047                            86.82
Musik / Tanz                   919                            76.20
Kunst / Kultur                 798                            66.17
Literatur / Medien             736                            61.03
Gesellschaft / Partizipation   643                            53.32
Berufsorientierung             636                            52.74
Naturwissenschaft / Technik    422                            34.99
Sprachen                       421                            34.91
Handwerk                       370                            30.68
Umwelt                         325                            26.95

Um zu errechen wie viele Schulen für eine jeweilige Kategorie Aktivitäten anbieten, nutzen wir die Pivot-Funktion. *(pivot_table())*. Mit der Funktion gruppieren wir die Aktivitäten nach ihren Kategorien. Teilt man diese Werte (hier unter id) durch die Gesamtanzahl der Schulen in Sachsen, erhalten wir den prozentualen Anteil an Schulen die Aktivitäten in einer Kategorie anbieten.


### 2. Auswertung der Schulpartnerschaften in Sachsen

In [17]:
sachsen_df_partner.head(2)

ID                                             Schule  \
0  SN-4220216-0  Sonderpädagogisches Förderzentrum  "Johannes T...   
1  SN-4220216-0  Sonderpädagogisches Förderzentrum  "Johannes T...   

               Type_grob                              partner_name  \
0  Gemeinnütziger Akteur   Kinder-, Jugend- und Familienhilfe e.V.   
1             Unbestimmt  diverse Mittelschulen der Stadt Chemnitz   

                  school_type  
0  Schule für Erziehungshilfe  
1  Schule für Erziehungshilfe

Auch hier ist es zunächst sinnvoll,einen Blick in die Daten zu werfen.

In [32]:
schulkategorien = {'Grundschule' : [' Grundschule',
                                     'Grundschule'],
                    'Sekundarschule' : ['Abendmittelschule',
                                        'Mittelschule',
                                        'Schulversuch Gemeinschaftsschule',
                                        ' Mittelschule'],
                    'Berufliche Schule' : ['Berufsfachschule',
                                           'Berufsschule',
                                           'Fachschule',
                                           ' Fachoberschule',
                                           'Fachoberschule',
                                           ],
                    'Förderschulen' : ['Berufsbildende Förderschule',
                                       'Schule für geistig Behinderte',
                                       'Schule zur Lernförderung',
                                       'Sprachheilschule',
                                       'Schule für Blinde und Sehbehinderte',
                                       'Schule für Erziehungshilfe',
                                       'Schule für geistig Behinderte',
                                       'Schule für Hörgeschädigte',
                                       'Schule für Körperbehinderte',
                                       'Schule zur Lernförderung',
                                       'Sprachheilschule'],
                    'Sonstige' :['Klinik- und Krankenhausschule',
                                 ' Klinik- und Krankenhausschule',
                                 'Kolleg'],
                   
                   'Gymnasium' : ['Abendgymnasium',
                                  'Berufliches Gymnasium',
                                  'Gymnasium']
                   
                   }

Um die unterschiedlichen Schultypen zu vereinheitlichen, ordnen wir diese in 6 Kategorien ein: Grundschule, Sekundarschule, Berufliche Schule, Förderschulen, Gymnasium, Sonstige. Dazu erstellen wir das Dictionary: Schulkategorien.

In [30]:
def check_up_Schultyp(schooltype):
    for key, value in schulkategorien.items():
        for a in value:
            if a in str(schooltype):
                return(key)

In [25]:
sachsen_df_partner = pd.melt(sachsen_df_partner, id_vars=['Type_grob','partner_name','Schule','ID'], 
                  var_name="Schulkategorien", value_name="Schultyp")

In Sachsen kann eine Schule aus mehreren Schulen bestehen. Schule xy ist beispielsweise ein Gymansium und eine Berufliche Schule zugleich. Mit der Funktion *'melt()'* transformieren wir das Datenset so, dass für jeden Schultyp ein Zeileneintrag entsteht.

In [34]:
# Hier mappen wir die einzelnen Schultypen auf die 6 Schultypkategorien und säubern leere Zeilen aus.
sachsen_df_partner['Type_grob'] = sachsen_df_partner['Schule'].apply(check_up_Schultyp)
# Einträge mit leeren Zellen werden hier ausgefiltert
sachsen_df_partner = sachsen_df_partner.dropna(subset=['Schule'])


In [35]:
df_sachsen.Schulname.count()

6025

Indem wir alle Einträge in dem Datenset zählen, sehen wir wie viele Partnerschaften Schulen in Sachsen eingehen. 

In [38]:

df_Sachsen2 = sachsen_df_partner.drop_duplicates(subset=['Schultyp', 'ID'])
df_Sachsen2.Schultyp.count()

945

Indem wir nur Reihen zählen indenen der Dienststellenschlüssel und der Schultyp einzigartig sind, erhalten wir die Anzahl der Schulen die an das Statistische Landesamt Sachsen kommuniziert haben.

In [44]:
Typ_Partner = pd.DataFrame(sachsen_df_partner.Type_grob.value_counts())
Typ_Partner = pd.DataFrame(pd.pivot_table(sachsen_df_partner,index=['Type_grob'],values='partner_name', aggfunc='count'))
Typ_Partner['Verteilung'] = (Typ_Partner / sum(Typ_Partner.partner_name)).round(4)
Typ_Partner

partner_name  Verteilung
Type_grob                                  
Berufliche Schule             1      0.0003
Förderschulen               689      0.2175
Grundschule                1660      0.5240
Gymnasium                   806      0.2544
Sonstige                     12      0.0038

Um zu sehen wie viele Partnerschaften in eine Akteurskategorie fallen und die Anteile der Kategorien im Verhältnis zueinander sind nutzen wir erneut die Pivotfunktion. Dadurch Gruppieren wir die Partnerschaften nach den jeweiligen Kategorien. Teilt man die Werte durch die Gesamtzahl aller Partnerschaften erhaält man die Verteilung. 

##### Wie viele Akteure für die jeweilige Kategorie kommen auf  eine Schule in Sachsen?

In [50]:
#Schutyp_unique = pd.pivot_table(sachsen_df_partner,index=['school_type'],values='partner_name',columns='Type_grob', aggfunc='count')
#Schutyp_unique = Schutyp_unique.fillna(0)
sachsen_df_partner

Type_grob                                       partner_name  \
0      Förderschulen            Kinder-, Jugend- und Familienhilfe e.V.   
1      Förderschulen           diverse Mittelschulen der Stadt Chemnitz   
2      Förderschulen   Solaris Förderzentrum für Jugend und Umwelt GmbH   
3      Förderschulen              Jugendamt Stadt Plauen, Vogtlandkreis   
4      Förderschulen                        Daetz- Zentrum Lichtenstein   
5      Förderschulen                                            SV 1860   
6      Förderschulen  Verein zur Förderung von Jugend und Sozialarbe...   
7      Förderschulen                       Verein für Schulsozialarbeit   
8      Förderschulen                     Kinder- und Jugendheim der AWO   
9      Förderschulen  Pirnaer Werkstätten Anerkannte Werkstatt für B...   
10     Förderschulen                      Musikschule Sächs.Schweiz e.V   
11     Förderschulen                                Universität Leipzig   
12     Förderschulen  sieben verschiedenen mitteständischen Betriebe...   
13     Förderschulen                   AWO Sonnenstein Werkstätten GmbH   
14              None                                               SAEK   
15              None                                               SAEK   
16              None                             Mittelschule Bernstadt   
17              None                             Mittelschule Bernstadt   
18              None                        Hochschule Zittau / Görlitz   
19              None                        Hochschule Zittau / Görlitz   
20              None                  41 Unternehmen der Region Görlitz   
21              None                  41 Unternehmen der Region Görlitz   
22              None  Ausbildungsring der Metall- und Elektroindustr...   
23              None  Ausbildungsring der Metall- und Elektroindustr...   
24              None                                Lausitz Matrix e.V.   
25              None                                Lausitz Matrix e.V.   
26              None                                               DLRG   
27              None                                               DLRG   
28              None                               Internationaler Bund   
29              None                               Internationaler Bund   
...              ...                                                ...   
10670    Grundschule                        Kita FlohkisteBerggießhübel   
10671    Grundschule                     Musikschule Sächsische Schweiz   
10672    Grundschule         Förderverein der Grundschule Berggießhübel   
10673    Grundschule                                   Hort Riffpiraten   
10674    Grundschule                       Median Kliniken Bad Lausick#   
10675    Grundschule                                     MS Bad Lausick   
10676    Grundschule                                  Riff Bad Lausick#   
10677    Grundschule                            Sportverein Bad Lausick   
10678    Grundschule                                      SSV Altenberg   
10679    Grundschule                                         TU Dresden   
10680    Grundschule                                         TU Dresden   
10681    Grundschule                             Stadtbücherei Auerbach   
10682    Grundschule                         Judoverein IPPON Rodewisch   
10683    Grundschule  Bibliothek Auerbach,Gemeinnütziges Schulungsze...   
10684    Grundschule                                Bibliothek Auerbach   
10685    Grundschule      Gemeinnütziges Schulungszentrum Witt Auerbach   
10686    Grundschule                                    Stadtbibliothek   
10687    Grundschule                    Jugendhaus der AWO in Bad Düben   
10688    Grundschule               KiTa \Märchenland\ und \Spatzenhaus\   
10689    Grundschule                         Kreismusikschule Delitzsch   
10690    Grundschule                               Oberschule Bad Düben   
10691    Grundschule                

In [18]:
Schutyp_unique = pd.DataFrame(Schutyp_unique)

Wir nutzen wieder die Pivotfunktion. Wir zählen  allerdings nur die Akteurskategorien die auf einen Schultyp kommen.

In [19]:
dftyp = df_Sachsen.drop_duplicates(subset=['Dienststellenschlüssel','Schultyp'])
dftyp = pd.DataFrame(dftyp.Schultypen_grob.value_counts())
dftyp

Schultypen_grob
Grundschule                    448
Sekundarschule                 259
Gymnasium                      148
Berufliche Schule              129
Förderschulen                  120
Sonstige                         2

Hier errechnen wir die Anzahl der Schulen in Sachsen für jeden Schultyp.

In [20]:
Schutyp_unique['Schulanzahl'] = dftyp

Anschließend fügen wir die Anzahl der Schulen pro Schultyp an das zuvorerstellte Dataframe mit der Anzahl der Partnerschaften pro Schultyp.

In [21]:
Schutyp_unique[['Gemeinnütziger Akteur','Partnerschule','Religiöse Einrichtung', 'Unbestimmt', 'Verband / Kammer / Innung / Gewerkschaft', 'Wirtschaftsakteur', 'Öffentliche Infrastruktur']].div(Schutyp_unique.Schulanzahl, axis=0)

Type_broad         Gemeinnütziger Akteur  Partnerschule  \
Schultypen_grob                                           
Berufliche Schule               1.224806       0.961240   
Förderschulen                   2.466667       0.416667   
Grundschule                     1.814732       0.357143   
Gymnasium                       1.770270       0.317568   
Sekundarschule                  3.061776       0.212355   
Sonstige                        3.000000       1.500000   

Type_broad         Religiöse Einrichtung  Unbestimmt  \
Schultypen_grob                                        
Berufliche Schule               0.046512    0.294574   
Förderschulen                   0.341667    0.616667   
Grundschule                     0.062500    0.154018   
Gymnasium                       0.108108    0.310811   
Sekundarschule                  0.158301    0.501931   
Sonstige                        0.000000    0.000000   

Type_broad         Verband / Kammer / Innung / Gewerkschaft  \
Schultypen_grob                                               
Berufliche Schule                                  1.480620   
Förderschulen                                      0.216667   
Grundschule                                        0.042411   
Gymnasium                                          0.405405   
Sekundarschule                                     0.177606   
Sonstige                                           0.500000   

Type_broad         Wirtschaftsakteur  Öffentliche Infrastruktur  
Schultypen_grob                                                  
Berufliche Schule           0.744186                   1.976744  
Förderschulen               1.058333                   2.150000  
Grundschule                 0.214286                   2.160714  
Gymnasium                   1.054054                   3.074324  
Sekundarschule              1.559846                   2.907336  
Sonstige                    0.500000                   2.000000

Anschließend teilen wir die Anzahl der Partnerschaften pro Akteurskategorie durch die jeweilige Anzahl der Schulen pro Schultyp. Die obenstehende Tabelle zeigt also für jeden Schultyp wie viele Partnerschaften sie durchschnittlich pro Akteurskategorie eingeht.